File: packages/clinical-contracts/src/rule4/phase/durationCalendar.ts
const ISO_DATE = /^(\d{4})-(\d{2})-(\d{2})$/;

export type ParsedIsoDate = { y: number; m: number; d: number };

export function parseIsoDateOnly(value: string | null | undefined): ParsedIsoDate | null {
  if (value == null || typeof value !== 'string') {
    return null;
  }
  const trimmed = value.trim();
  const m = ISO_DATE.exec(trimmed);
  if (!m) {
    return null;
  }
  const y = Number(m[1]);
  const mo = Number(m[2]);
  const d = Number(m[3]);
  if (!isValidCalendarDate(y, mo, d)) {
    return null;
  }
  return { y, m: mo, d };
}

function isValidCalendarDate(y: number, m: number, d: number): boolean {
  if (m < 1 || m > 12 || d < 1) {
    return false;
  }
  const dim = daysInMonth(y, m);
  return d <= dim;
}

function daysInMonth(y: number, m: number): number {
  if (m === 2) {
    return isLeapYear(y) ? 29 : 28;
  }
  if ([4, 6, 9, 11].includes(m)) {
    return 30;
  }
  return 31;
}

function isLeapYear(y: number): boolean {
  if (y % 400 === 0) {
    return true;
  }
  if (y % 100 === 0) {
    return false;
  }
  return y % 4 === 0;
}

/** UTC calendar day number for stable arithmetic (no timezone). */
function dayNumber({ y, m, d }: ParsedIsoDate): number {
  let y2 = y;
  let m2 = m;
  if (m2 <= 2) {
    y2 -= 1;
    m2 += 12;
  }
  const era = Math.floor(y2 / 400);
  const yoe = y2 - era * 400;
  const doy = Math.floor((153 * (m2 - 3) + 2) / 5) + d - 1;
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy;
  return era * 146097 + doe - 719468;
}

/**
 * Inclusive duration: same day → 1; assessment before onset → null (invalid).
 */
export function inclusiveDurationDays(
  onset: string | null | undefined,
  assessment: string | null | undefined,
): { days: number | null; invalid: boolean } {
  const o = parseIsoDateOnly(onset);
  const a = parseIsoDateOnly(assessment);
  if (!o || !a) {
    return { days: null, invalid: false };
  }
  const diff = dayNumber(a) - dayNumber(o);
  if (diff < 0) {
    return { days: null, invalid: true };
  }
  return { days: diff + 1, invalid: false };
}

export function isValidPositiveIntegerDuration(raw: unknown): raw is number {
  return typeof raw === 'number' && Number.isInteger(raw) && raw > 0;
}

File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/phase/duration_calendar.py
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date

ISO_DATE = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")


@dataclass(frozen=True)
class ParsedIsoDate:
    y: int
    m: int
    d: int


def parse_iso_date_only(value: str | None) -> ParsedIsoDate | None:
    if value is None or not isinstance(value, str):
        return None
    trimmed = value.strip()
    m = ISO_DATE.match(trimmed)
    if not m:
        return None
    y, mo, d = int(m.group(1)), int(m.group(2)), int(m.group(3))
    try:
        date(y, mo, d)
    except ValueError:
        return None
    return ParsedIsoDate(y=y, m=mo, d=d)


def _day_number(parts: ParsedIsoDate) -> int:
    """UTC calendar day number for stable arithmetic (matches TS dayNumber)."""
    y2, m2 = parts.y, parts.m
    if m2 <= 2:
        y2 -= 1
        m2 += 12
    era = y2 // 400
    yoe = y2 - era * 400
    doy = (153 * (m2 - 3) + 2) // 5 + parts.d - 1
    doe = yoe * 365 + yoe // 4 - yoe // 100 + doy
    return era * 146097 + doe - 719468


def inclusive_duration_days(
    onset: str | None, assessment: str | None
) -> dict:
    o = parse_iso_date_only(onset)
    a = parse_iso_date_only(assessment)
    if not o or not a:
        return {"days": None, "invalid": False}
    diff = _day_number(a) - _day_number(o)
    if diff < 0:
        return {"days": None, "invalid": True}
    return {"days": diff + 1, "invalid": False}


def is_valid_positive_integer_duration(raw: object) -> bool:
    return isinstance(raw, int) and not isinstance(raw, bool) and raw > 0
